README.md

```diff
@@ -1319,6 +1186,157 @@ should be added to `globalExternalImplsMap` in
+`atomicExchange`, which also returns the old value of the reference:
+
+```
+utest atomicExchange a 2 with 1 in
+```
+
+Finally, for integer references, we can use `atomicFetchAndAdd` to increase or
+decrease the value of the reference. The function returns the old value of the
+reference:
+
+```
+utest atomicFetchAndAdd a 1 with 2 in
+-- Current value is now 3
+utest atomicFetchAndAdd a (subi 0 45) with 3 in
+-- Current value is now -42
+```
+
+#### Multi-Threaded Execution
+
+Functions for handling threads are provided in
+[multicore/threads.mc](stdlib/multicore/threads.mc).
+The following example program spawns 10 threads that compete for printing their
+IDs:
+
+```
+include "string.mc"
+include "multicore/thread.mc"
+mexpr
+let place = atomicMake 1 in
+let threads = create 10 (lam. threadSpawn (lam.
+  print (join
+    [int2string (atomicFetchAndAdd place 1)
+    , ": thread ID "
+    , int2string (threadID2int (threadSelf ()))
+    , "\n"
+    ]))
+) in
+map threadJoin threads
+```
+
+where `threadSpawn` takes a function of type `Unit -> a` as argument,
+`threadSelf` returns the ID of the current thread, and `threadID2int` converts a
+thread ID to a unique integer. Note that `threadJoin` must be called once for
+each call to `threadSpawn`. The output of the above program might be:
+
+```
+1: thread ID 1
+2: thread ID 2
+3: thread ID 129
+4: thread ID 130
+5: thread ID 3
+6: thread ID 257
+7: thread ID 258
+8: thread ID 131
+9: thread ID 385
+10: thread ID 386
+```
+
+However, the values and order of the thread IDs might be different over
+different runs.
+
+To control the execution order of threads, some form of thread synchronization
+is necessary. This can be done either using atomic references, or using the
+functions `threadCriticalSection`, `threadWait` and `threadNotify`, or both.
+In the following example:
+
+```
+let inCriticalSection = atomicMake false in
+let afterWait = atomicMake false in
+
+let t = threadSpawn (lam.
+  threadCriticalSection (
+    lam.
+      atomicExchange inCriticalSection true;
+      threadWait ();
+      atomicExchange afterWait true
+  )
+) in
+```
+
+the thread `t` enters a critical section, where it first atomically sets the
+`inCriticalSection` flag to true, before calling `threadWait`, which will block
+the execution until another thread calls `threadNotify`. In the following, the
+main thread first waits for `t` to set the `inCriticalSection` flag, before
+making a call to `threadNotify`:
+
+```
+recursive let waitForFlag = lam flag.
+  match atomicGet flag with true then ()
+  else waitForFlag flag
+in
+waitForFlag inCriticalSection;
+threadNotify (threadGetID t);
+```
+
+The `threadNotify` will block the execution of the main thread until `t` has
+finished its critical section, which means that we know that the flag
+`afterWait` must now be true:
+
+```
+utest atomicGet afterWait with true in
+-- Don't forget to clean up!
+threadJoin t;
+```
+
```

Comment:
The `threadWait`, `threadNotify` and `threadCriticalSection` have been removed (they were removed from ocaml multicore API and I'm currently replacing them with something else (probably mutexes and condition varibles). It would be great if you could just remove this chunk of text for now.

Reply:
Done!